### ISSUE TITLE
Question regarding removing <start> from the targets during training

### ISSUE THREAD

**Issue by nilinykh** [sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning#94]
In `train.py`, it says the following:

     # Since we decoded starting with <start>, the targets are all words after <start>, up to <end>
     targets = caps_sorted[:, 1:]

What is meant here by `decoded starting with <start>`? 

The reason why I am asking is because, let's say, I have `scores.data` (`data` because of PyTorch latest version), which is a matrix of projected words, its shape is 5 x 20 x 10500 (for example, so I have `batch_size x seq_length x num_dim (vocab_size)`), but if I execute the line above, my targets will have a different shape: 5 x 19, and I would get an error in my criterion (score and target should have the same 1st dimension).

I have looked at the code in `models.py`, thinking that maybe it includes some manipulation with scores (so that when they are passed as predictions to `train.py` they definitely do not include <start> token to match the targets' shape), but found no line which does that.

So, could you please elaborate on why do we need to execute this line of code (num 175 in `train.py`)? Thank you.

**Comment by kmario23** [comment]
Hi,
  it seems you're missing the  (`pack_padded_sequence(...)`) transformation that happens between the lines

```
# Since we decoded starting with <start>, the targets are ....
 targets = caps_sorted[:, 1:]
```

> pack_padded_sequence(...)

 ```
# Calculate loss
loss = criterion(scores, targets)
```

So, after you remove the paddings etc.  using `pack_padded_sequence(...)`, both `scores` and `targets` tensor would have the same first dimension and broadcasting happens in the second dimension of `targets` tensor (which is a just a vector / 1D tensor in our case here). 

For a concrete example of the inputs that are passed to CE loss, see `examples` in [`nn.CrossEntropyLoss()`](https://pytorch.org/docs/stable/nn.html#torch.nn.CrossEntropyLoss)

Also, note that, though each entry (i.e. row) in the `scores` matrix is a score over the whole vocabulary, it's not a probability distribution (i.e. it doesn't sum to 1). The reason is that taking `log_softmax` at this stage is expensive and not needed since this can be considered as a normalization and it doesn't affect the training procedure. However, when we want to generate a caption (i.e. a word at each timestep), then we must apply the softmax operation as done in [`caption.py`](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning/blob/master/caption.py#L94)

**Comment by nilinykh** [comment]
Hey,

thank you for your quick reply!

I actually do have `pack_padded_sequence(...)` in my code, and criterions does not complan if I pass original scores and targets. However, if I do `targets = caps_sorted[:, 1:]` in particular, this will through an error, since targets' shape now is not identical to the scores' shape.

The question is the following: does it mean that in `scores` we don't decode the first symbol, e.g. <start>?

**Comment by nilinykh** [comment]
@kmario23 ok, I think I found the problem (it was a bug in my code).
However, it is still unclear to me why do we take as targets all words after <start>, but not <start> itself.

**Comment by kmario23** [comment]
@nilinykh good! The reason why we don't consider `<start>` in the `targets` tensor is the following: at the time step `t=0`, we give two inputs to the decoder: (i) an initial hidden state (here that'd be the encoder output) (ii) the `<start>` token (i.e. its corresponding embedding from the embedding matrix.

Given these two inputs, the decoder (i.e. LSTM) would do its magic and predict the next probable token. This token should be the output that should be used for computing the `loss`; But, here we don't do that since we use teacher forcing. That is to say, we ignore the output generated by the LSTM and instead use the ground truth output that we already have.

For a concrete example, please see [seq2seq translation tutorial](https://pytorch.org/tutorials/intermediate/seq2seq_translation_tutorial.html)

**Comment by nilinykh** [comment]
Yes, thank you for your detailed explanations! They are very helpful indeed.
I believe that the issue is resolved.
